test/common/router/scoped_rds_test.cc

```diff
@@ -930,7 +971,516 @@ route_configuration_name: foo_routes
+                ->name(),
+            "");
+  EXPECT_THAT(getScopedRdsProvider()->config<ScopedConfigImpl>()->getRouteConfig(
+                  TestRequestHeaderMapImpl{{"Addr", "x-foo-key;x-bar-key"}}),
+              IsNull());
+  EXPECT_EQ(1UL, server_factory_context_.scope_
+                     .gauge("foo.scoped_rds.foo_scoped_routes.active_scopes",
+                            Stats::Gauge::ImportMode::Accumulate)
+                     .value());
+
+  ScopeKeyPtr scope_key = getScopedRdsProvider()->config<ScopedConfigImpl>()->computeScopeKey(
+      TestRequestHeaderMapImpl{{"Addr", "x-foo-key;x-bar-key"}});
+  EXPECT_CALL(event_dispatcher_, post(_)).Times(1);
+  std::function<void(bool)> route_config_updated_cb = [](bool) {};
+  getScopedRdsProvider()->onDemandRdsUpdate(std::move(scope_key), event_dispatcher_,
+                                            std::move(route_config_updated_cb));
+  // After on demand request, push rds update, both scopes should find the route configuration.
+  pushRdsConfig({"foo_routes"}, "111");
+  EXPECT_EQ(getScopedRdsProvider()
+                ->config<ScopedConfigImpl>()
+                ->getRouteConfig(TestRequestHeaderMapImpl{{"Addr", "x-foo-key;x-foo-key"}})
+                ->name(),
+            "foo_routes");
+  EXPECT_EQ(getScopedRdsProvider()
+                ->config<ScopedConfigImpl>()
+                ->getRouteConfig(TestRequestHeaderMapImpl{{"Addr", "x-foo-key;x-bar-key"}})
+                ->name(),
+            "foo_routes");
+  // Now we have 1 active on demand scope and 1 eager loading scope.
+  EXPECT_EQ(2UL, server_factory_context_.scope_
+                     .gauge("foo.scoped_rds.foo_scoped_routes.all_scopes",
+                            Stats::Gauge::ImportMode::Accumulate)
+                     .value());
+  EXPECT_EQ(2UL, server_factory_context_.scope_
+                     .gauge("foo.scoped_rds.foo_scoped_routes.active_scopes",
+                            Stats::Gauge::ImportMode::Accumulate)
+                     .value());
+  EXPECT_EQ(1UL, server_factory_context_.scope_
+                     .gauge("foo.scoped_rds.foo_scoped_routes.on_demand_scopes",
+                            Stats::Gauge::ImportMode::Accumulate)
+                     .value());
+}
+
+TEST_F(ScopedRdsTest, PushRdsBeforeOndemandRequest) {
+  setup();
+  init_watcher_.expectReady();
+  // Scope should be loaded eagerly by default.
+  const std::string config_yaml = R"EOF(
+name: foo_scope
+route_configuration_name: foo_routes
+key:
+  fragments:
+    - string_key: x-foo-key
+)EOF";
+  const auto eager_resource = parseScopedRouteConfigurationFromYaml(config_yaml);
+  // On demand scope should be loaded lazily.
+  const std::string config_yaml2 = R"EOF(
+name: foo_scope2
+route_configuration_name: foo_routes
+on_demand: true
+key:
+  fragments:
+    - string_key: x-bar-key
+)EOF";
+  const auto lazy_resource = parseScopedRouteConfigurationFromYaml(config_yaml2);
+
```

Comment:
I don't know if it's a reasonable ask or not to refactor some of these tests, but there is a fair bit of boiler plate and the logic of the test is lost somewhat due to that. I wouldn't do any gymnastics to avoid boiler plate in tests but a quick ask on whether it's possible to reduce some of the more verbose bits?

Reply:
Sure, we can wrap it inside a function that can be reused to make it less verbose.